fix(#1101): Make JsonTextMessageValidator ignore the order of JsonArrays

novarx opened the pull request:
See #1101 

@bbortt looking forward to your review / opinion =)

novarx commented on validation/citrus-validation-json/src/test/java/org/citrusframework/validation/json/JsonElementValidatorTest.java:
```diff
@@ -0,0 +1,272 @@
+
+        fixture = new JsonElementValidator(NOT_STRICT, context, jsonAssertion.ignoreExpressions);
+
+        // TODO that shouldn't fail, should it?
```
On second thought... 

`$.index['anything'][*]` should ignore all the children of `anything`, but the entry itself should be there nonetheless. Will update the test accordingly.

bbortt replied:
I agree.

novarx commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonTextMessageValidator.java:
```diff
@@ -56,190 +46,74 @@
      * Validates JSON text with comparison to expected control JSON object.
      * JSON entries can be ignored with ignore placeholder.
+     * <p>
+     * TODO is that method still needed?
```
It's nowhere used, but could still be a braking change - any thoughts?

novarx replied:
Added `@deprecated use .. instead`

novarx commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonTextMessageValidator.java:
```diff
@@ -56,190 +46,74 @@
      * Checks if given element node is either on ignore list or
      * contains @ignore@ tag inside control message
+     * <p>
+     * TODO is that method still needed?
```
It's nowhere used, but could still be a braking change - any thoughts?

bbortt replied:
I think just deprecate it for removal and leave an `@deprecated use .. instead` comment. this way we can remove it within one of the next few releases.

bbortt commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonElementValidator.java:
```diff
@@ -0,0 +1,135 @@
+package org.citrusframework.validation.json;
```
please add the missing copyright for 2024. also, all touched classes should be updated to copyright years "up to 2024" (`...-2024`).

bbortt commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonElementValidator.java:
```diff
@@ -0,0 +1,135 @@
+            validateSameSize(control.getJsonPath(), arrayControl.expected, arrayControl.actual);
+        }
+        for (int i = 0; i < arrayControl.expected.size(); i++) {
+            if (!isAnyValidItemInActualArray(validator, arrayControl, i)) {
```
so just to be clear; strict mode does **not** strictly validate the order of elements - right? I think that is kinda okay, since JSON in general has no such restriction. on the other hand, someone might want to test it e.g. when the return something ordered (paginated) from their APIs. we would not allow them to do so with the current implementation. any thoughts?

novarx replied:
Yeah that's true, I also thought about that...

Options are:
1. Always ignore order
2. Ignore order if `strict = false` and enforce order if `strict = true`
3. Make `JsonElementValidator` more configurable with properties 
4. Make `JsonElementValidator` more configurable by easily allowing overriding of the element-validations (in the refactoring process, I thougt of a [Chain of Responsibility](https://refactoring.guru/design-patterns/chain-of-responsibility) where each link is one element validation... but found that too sophisticated for the moment)

(I fancy Option 1 or 2)

bbortt replied:
so, option one is now implemented. I think that is acceptable. if someone really wants to check the order, he can easily implement a custom validator.

novarx replied:
seems to be a problem - addressed in https://github.com/citrusframework/citrus/pull/1155

bbortt commented on validation/citrus-validation-json/src/test/java/org/citrusframework/validation/json/JsonElementValidatorTest.java:
```diff
@@ -0,0 +1,272 @@
+        return JsonElementValidatorItem.parseJson(DEFAULT_PERMISSIVE_MODE, jsonAssertion.actual, jsonAssertion.expected);
+    }
+
+    public record JsonAssertion(
```
doesn't need to be `public`, right?

bbortt reviewed (changes requested):
couldn't find much, that is a very good addition and refactoring. thank you!

novarx commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonElementValidatorItem.java:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * @return {@link JsonElementValidatorItem#actual} as string or null
+     */
+    public String actualString() {
```
changed it to `actualAsStringOrNull()`

christophd commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonTextMessageValidator.java:
```diff
@@ -56,225 +46,85 @@
      * @param readContext
      * @return
      */
+    @Deprecated(forRemoval = true)
```
If this method is not called anymore in the new validation logic we should rather remove it now. This is because somebody might have customized this method and if it is not called anymore it may be more confusing. Rather break the API and let users adjust accordingly.

novarx replied:
Removed that method completely 👍

christophd commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonTextMessageValidator.java:
```diff
@@ -56,225 +46,85 @@
+     * @param readContext       the JSONPath read context.
      */
-    @SuppressWarnings("rawtypes")
+    @Deprecated(forRemoval = true)
```
If we deprecate we should call the method as part of the new logic. Otherwise when the method is not called I would rather remove and break the API. Then users that have added customizations to this method get noticed when updating and can adjust accordingly.

@bbortt WDYT?

bbortt replied:
that argumentation is valid. it support it 😉

christophd commented on validation/citrus-validation-json/src/main/java/org/citrusframework/validation/json/JsonTextMessageValidator.java:
```diff
@@ -56,225 +46,85 @@
+            throw new ValidationException("Validation failed - expected message contents, but received empty message!");
         }
 
+        new JsonElementValidator(strict, context, validationContext.getIgnoreExpressions()).validate(
```
How about sharing the instance of JsonElementValidator in this class as a private member. Users may want to set a customized implementation of JsonElementValidator then.

We can pass the parameters to the JsonElementValidator#validate method then (e.g. context and validationContext) so JsonElementValidator is stateless

novarx replied:
I've introduced a `JsonElementValidator.Provider` to configure `JsonTextMessageValidator` with a custom element validator - hope that's fine?

christophd reviewed (commented):
Many thanks for this contribution! I have some comments on the deprecated methods. Please have a look and let me know if you have questions.

Special thanks for refactoring the unit tests! I can not verify that all test case variants are covered as before (or even more than that) but I trust you 😃. Thanks a lot!

novarx commented:
I accidentally closed the PR, didn't mean to - sorry 😅

novarx commented on validation/citrus-validation-json/pom.xml:
```diff
@@ -72,5 +72,11 @@
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <version>3.25.1</version>
```
Done!

bbortt reviewed (changes requested):
almost.. - sorry 😅

novarx commented:
Thanks for the reviews!

@christophd I've double-checked, that I only deleted testcases in `JsonTextMessageValidatorTest` that I moved to `JsonElementValidatorTest`. Further, I've added some array-ordering-tests.

bbortt commented:
finally.. god, that build 🤣
